zkevm-circuits/src/util/cell_manager.rs

```diff
@@ -1,8 +1,269 @@
+            column_expression: meta.query_advice(column, Rotation::cur()),
+            column,
+            column_idx,
+            rotation,
+            cell_column_index: column_idx,
+        }
+    }
+}
+
+impl<F: Field> Cell<F> {
+    pub fn new_from_cs(
+        meta: &mut ConstraintSystem<F>,
+        column: Column<Advice>,
+        column_idx: usize,
+        rotation: usize,
+    ) -> Cell<F> {
+        query_expression(meta, |meta| Cell::new(meta, column, column_idx, rotation))
+    }
+
+    pub(crate) fn assign(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        value: Value<F>,
+    ) -> Result<AssignedCell<F, F>, Error> {
+        region.assign_advice(
+            || {
+                format!(
+                    "Cell column: {:?} and rotation: {}",
+                    self.column, self.rotation
+                )
+            },
+            self.column,
+            offset + self.rotation,
+            || value,
+        )
+    }
+}
+
+impl<F: Field> Expr<F> for Cell<F> {
+    fn expr(&self) -> Expression<F> {
+        self.expression.clone()
+    }
+}
+
+impl<F: Field> Expr<F> for &Cell<F> {
+    fn expr(&self) -> Expression<F> {
+        self.expression.clone()
+    }
+}
+
+#[derive(Debug, Clone)]
+/// CellColumn
+pub(crate) struct CellColumn {
+    pub advice: Column<Advice>,
+    pub cell_type: CellType,
+    pub idx: usize,
+}
+
+impl CellColumn {
+    pub fn new(advice: Column<Advice>, cell_type: CellType, idx: usize) -> CellColumn {
+        CellColumn {
+            advice,
+            cell_type,
+            idx,
+        }
+    }
+
+    pub fn expr<F: Field>(&self, meta: &mut ConstraintSystem<F>) -> Expression<F> {
+        query_expression(meta, |meta| meta.query_advice(self.advice, Rotation::cur()))
+    }
+}
+
+/// CellManagerStrategy
+pub(crate) trait CellManagerStrategy {
+    type Stats;
+
+    /// on_creation
+    fn on_creation(&mut self, columns: &mut CellManagerColumns);
+    /// query_cell
+    fn query_cell<F: Field>(
+        &mut self,
+        columns: &mut CellManagerColumns,
+        meta: &mut ConstraintSystem<F>,
+        cell_type: CellType,
+    ) -> Cell<F>;
+    /// get_height
     fn get_height(&self) -> usize;
+
+    fn get_stats(&self, columns: &CellManagerColumns) -> Self::Stats;
+}
+
+#[derive(Default, Debug, Clone)]
+pub(crate) struct CellManagerColumns {
+    columns: HashMap<CellType, Vec<CellColumn>>,
+    columns_list: Vec<CellColumn>,
+}
+
+impl CellManagerColumns {
+    pub fn add_column(&mut self, cell_type: CellType, column: Column<Advice>) {
+        let idx = self.columns_list.len();
+        let cell_column = CellColumn::new(column, cell_type, idx);
+        self.columns_list.push(cell_column.clone());
+        if let Some(columns) = self.columns.get_mut(&cell_type) {
+            columns.push(cell_column);
+        } else {
+            self.columns.insert(cell_type, vec![cell_column]);
+        }
```

Comment:
This is a nitpick, but the `HashMap` has a nice API to do update/insert: https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry
So this could be:
```rust
self.columns.entry(&cell_type).and_modify(|columns| columns.push(cell_column)).or_insert(vec![cell_column]);
```

Reply:
Great!! Didn't know. I think it is good to say this things to spread the knowledge, not nitpick